File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters/Session/OriginState/RecordBasedOriginState.cs
// <copyright file="RecordBasedOriginState.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters
{
    using System.Collections.Generic;
    using System.Linq;
    using Meta;
    using Numbers;

    public abstract class RecordBasedOriginState
    {
        public abstract Strategy Strategy { get; }

        protected bool HasChanges => this.Record == null || this.ChangedRoleByRelationType?.Count > 0;

        protected abstract IEnumerable<IRoleType> RoleTypes { get; }

        protected abstract IRecord Record { get; }

        protected IRecord PreviousRecord { get; set; }

        public Dictionary<IRelationType, object> ChangedRoleByRelationType { get; protected set; }

        private Dictionary<IRelationType, object> PreviousChangedRoleByRelationType { get; set; }

        public object GetRole(IRoleType roleType)
        {
            if (this.ChangedRoleByRelationType != null &&
                this.ChangedRoleByRelationType.TryGetValue(roleType.RelationType, out var role))
            {
                return role;
            }

            return this.Record?.GetRole(roleType);
        }

        public void SetUnitRole(IRoleType roleType, object role) => this.SetChangedRole(roleType, role);

        public void SetCompositeRole(IRoleType roleType, long? role)
        {
            var previousRole = (long?)this.GetRole(roleType);

            if (previousRole == role)
            {
                return;
            }

            var associationType = roleType.AssociationType;

            if (associationType.IsOne && role.HasValue)
            {
                var previousAssociationObject =
                    this.Session.GetAssociation<IObject>(role.Value, associationType).FirstOrDefault();

                this.SetChangedRole(roleType, role);

                if (associationType.IsOne && previousAssociationObject != null)
                {
                    // OneToOne
                    previousAssociationObject?.Strategy.Set(roleType, null);
                }
            }
            else
            {
                this.SetChangedRole(roleType, role);
            }
        }

        public void AddCompositeRole(IRoleType roleType, long roleToAdd)
        {
            var previousRole = this.GetRole(roleType);

            if (this.Numbers.Contains(previousRole, roleToAdd))
            {
                return;
            }

            var role = this.Numbers.Add(previousRole, roleToAdd);

            this.SetChangedRole(roleType, role);

            var associationType = roleType.AssociationType;
            if (associationType.IsMany)
            {
                return;
            }

            // OneToMany
            var previousAssociationObject =
                this.Session.GetAssociation<IObject>(roleToAdd, associationType).FirstOrDefault();
            previousAssociationObject?.Strategy.Set(roleType, null);
        }

        public void RemoveCompositeRole(IRoleType roleType, long roleToRemove)
        {
            var previousRole = this.GetRole(roleType);

            if (!this.Numbers.Contains(previousRole, roleToRemove))
            {
                return;
            }

            var role = this.Numbers.Remove(previousRole, roleToRemove);

            this.SetChangedRole(roleType, role);
        }

        public void SetCompositesRole(IRoleType roleType, object role)
        {
            var previousRole = this.GetRole(roleType);

            this.SetChangedRole(roleType, role);

            var associationType = roleType.AssociationType;
            if (associationType.IsMany)
            {
                return;
            }

            // OneToMany
            var addedRoles = this.Numbers.Except(role, previousRole);
            foreach (var addedRole in this.Numbers.Enumerate(addedRoles))
            {
                var previousAssociationObject =
                    this.Session.GetAssociation<IObject>(addedRole, associationType).FirstOrDefault();
                previousAssociationObject?.Strategy.Set(roleType, null);
            }
        }

        public void Checkpoint(ChangeSet changeSet)
        {
            // Same record
            if (this.PreviousRecord == null || this.Record == null ||
                this.Record.Version == this.PreviousRecord.Version)
            {
                // No previous changed roles
                if (this.PreviousChangedRoleByRelationType == null)
                {
                    if (this.ChangedRoleByRelationType != null)
                    {
                        // Changed roles
                        foreach (var kvp in this.ChangedRoleByRelationType)
                        {
                            var relationType = kvp.Key;
                            var current = kvp.Value;
                            var previous = this.Record?.GetRole(relationType.RoleType);

                            changeSet.Diff(this.Strategy, relationType, current, previous);
                        }
                    }
                }
                // Previous changed roles
                else
                {
                    foreach (var kvp in this.ChangedRoleByRelationType)
                    {
                        var relationType = kvp.Key;
                        var current = kvp.Value;

                        _ = this.PreviousChangedRoleByRelationType.TryGetValue(relationType, out var previous);
                        changeSet.Diff(this.Strategy, relationType, current, previous);
                    }
                }
            }
            // Different record
            else
            {
                foreach (var roleType in this.RoleTypes)
                {
                    var relationType = roleType.RelationType;

                    object previous = null;
                    object current = null;

                    if (this.PreviousChangedRoleByRelationType?.TryGetValue(relationType, out previous) == true)
                    {
                        if (this.ChangedRoleByRelationType?.TryGetValue(relationType, out current) == true)
                        {
                            changeSet.Diff(this.Strategy, relationType, current, previous);
                        }
                        else
                        {
                            current = this.Record.GetRole(roleType);
                            changeSet.Diff(this.Strategy, relationType, current, previous);
                        }
                    }
                    else
                    {
                        previous = this.PreviousRecord?.GetRole(roleType);
                        if (this.ChangedRoleByRelationType?.TryGetValue(relationType, out current) == true)
                        {
                            changeSet.Diff(this.Strategy, relationType, current, previous);
                        }
                        else
                        {
                            current = this.Record.GetRole(roleType);
                            changeSet.Diff(this.Strategy, relationType, current, previous);
                        }
                    }
                }
            }

            this.PreviousRecord = this.Record;
            this.PreviousChangedRoleByRelationType = this.ChangedRoleByRelationType;
        }

        public bool IsAssociationForRole(IRoleType roleType, long forRole)
        {
            if (roleType.ObjectType.IsUnit)
            {
                return false;
            }

            var role = this.GetRole(roleType);

            if (roleType.IsOne)
            {
                return (long?)role == forRole;
            }

            return this.Numbers.Contains(role, forRole);
        }

        protected abstract void OnChange();

        private void SetChangedRole(IRoleType roleType, object role)
        {
            this.ChangedRoleByRelationType ??= new Dictionary<IRelationType, object>();
            this.ChangedRoleByRelationType[roleType.RelationType] = role;
            this.OnChange();
        }

        #region Proxy Properties

        protected long Id => this.Strategy.Id;

        protected IClass Class => this.Strategy.Class;

        protected Session Session => this.Strategy.Session;

        protected Workspace Workspace => this.Session.Workspace;

        private INumbers Numbers => this.Strategy.Session.Workspace.Numbers;

        #endregion
    }
}

File: typescript/libs/workspace/adapters/json/system/src/lib/session/originstate/RecordBasedOriginState.ts
import { Class, RelationType, RoleType } from '@allors/workspace/meta/system';
import { IRecord } from '../../IRecord';
import { ChangeSet } from '../ChangeSet';
import { Strategy } from '../Strategy';
import { Workspace } from '../../workspace/Workspace';
import { Session } from '../Session';
import { add, has, remove, difference, enumerate } from '../../collections/Numbers';

export abstract class RecordBasedOriginState {
  public abstract Strategy: Strategy;

  protected HasChanges(): boolean {
    return this.Record == null || this.ChangedRoleByRelationType?.size > 0;
  }

  protected abstract RoleTypes: RoleType[];

  protected abstract Record: IRecord;

  protected PreviousRecord: IRecord;

  public ChangedRoleByRelationType: Map<RelationType, any>;

  private PreviousChangedRoleByRelationType: Map<RelationType, any>;

  public getRole(roleType: RoleType): any {
    if (this.ChangedRoleByRelationType != null && this.ChangedRoleByRelationType.has(roleType.relationType)) {
      return this.ChangedRoleByRelationType.get(roleType.relationType);
    }

    return this.Record?.getRole(roleType);
  }

  public SetUnitRole(roleType: RoleType, role: any) {
    this.SetChangedRole(roleType, role);
  }

  public SetCompositeRole(roleType: RoleType, role?: number) {
    const previousRole = this.getRole(roleType) as number;
    if (previousRole == role) {
      return;
    }

    const associationType = roleType.associationType;
    if (associationType.isOne && role != null) {
      const previousAssociationObject = this.Session.getAssociation(role, associationType).FirstOrDefault();
      this.SetChangedRole(roleType, role);
      if (associationType.isOne && previousAssociationObject != null) {
        //  OneToOne
        previousAssociationObject?.Strategy.Set(roleType, null);
      }
    } else {
      this.SetChangedRole(roleType, role);
    }
  }

  public AddCompositeRole(roleType: RoleType, roleToAdd: number) {
    const previousRole = this.getRole(roleType);
    if (has(previousRole, roleToAdd)) {
      return;
    }

    const role = add(previousRole, roleToAdd);
    this.SetChangedRole(roleType, role);
    const associationType = roleType.associationType;
    if (associationType.isMany) {
      return;
    }

    //  OneToMany
    const previousAssociationObject = this.Session.getAssociation(roleToAdd, associationType).FirstOrDefault();
    previousAssociationObject?.Strategy.Set(roleType, null);
  }

  public RemoveCompositeRole(roleType: RoleType, roleToRemove: number) {
    const previousRole = this.getRole(roleType);
    if (!has(previousRole, roleToRemove)) {
      return;
    }

    const role = remove(previousRole, roleToRemove);
    this.SetChangedRole(roleType, role);
  }

  public SetCompositesRole(roleType: RoleType, role: any) {
    const previousRole = this.getRole(roleType);
    this.SetChangedRole(roleType, role);
    const associationType = roleType.associationType;
    if (associationType.isMany) {
      return;
    }

    //  OneToMany
    const addedRoles = difference(role, previousRole);
    for (const addedRole of enumerate(addedRoles)) {
      const previousAssociationObject = this.Session.getAssociation(addedRole, associationType).FirstOrDefault();
      previousAssociationObject?.Strategy.Set(roleType, null);
    }
  }

  public Checkpoint(changeSet: ChangeSet) {
    //  Same record
    if (this.PreviousRecord == null || this.Record == null || this.Record.version == this.PreviousRecord.version) {
      //  No previous changed roles
      if (this.PreviousChangedRoleByRelationType == null) {
        if (this.ChangedRoleByRelationType != null) {
          //  Changed roles
          this.ChangedRoleByRelationType.forEach((current, relationType) => {
            const previous = this.Record?.getRole(relationType.roleType);
            changeSet.Diff(this.Strategy, relationType, current, previous);
          });
        }
      }

      //  Previous changed roles
      this.ChangedRoleByRelationType.forEach((current, relationType) => {
        const previous = this.PreviousChangedRoleByRelationType.get(relationType);
        changeSet.Diff(this.Strategy, relationType, current, previous);
      });
    } else {
      //  Different record
      this.RoleTypes.forEach((roleType) => {
        const relationType = roleType.relationType;
        let previous = null;
        let current = null;
        if (this.PreviousChangedRoleByRelationType?.has(relationType)) {
          const previous = this.PreviousChangedRoleByRelationType.get(relationType);
          if (this.ChangedRoleByRelationType?.has(relationType)) {
            const current = this.ChangedRoleByRelationType.get(relationType);
            changeSet.Diff(this.Strategy, relationType, current, previous);
          } else {
            current = this.Record.getRole(roleType);
            changeSet.Diff(this.Strategy, relationType, current, previous);
          }
        } else {
          previous = this.PreviousRecord?.getRole(roleType);
          if (this.ChangedRoleByRelationType?.has(relationType)) {
            const current = this.ChangedRoleByRelationType?.get(relationType);
            changeSet.Diff(this.Strategy, relationType, current, previous);
          } else {
            current = this.Record.getRole(roleType);
            changeSet.Diff(this.Strategy, relationType, current, previous);
          }
        }
      });
    }

    this.PreviousRecord = this.Record;
    this.PreviousChangedRoleByRelationType = this.ChangedRoleByRelationType;
  }

  public IsAssociationForRole(roleType: RoleType, forRole: number): boolean {
    if (roleType.objectType.isUnit) {
      return false;
    }

    const role = this.getRole(roleType);
    if (roleType.isOne) {
      return role == forRole;
    }

    return has(role, forRole);
  }

  protected abstract OnChange();

  private SetChangedRole(roleType: RoleType, role: any) {
    this.ChangedRoleByRelationType ??= new Map<RelationType, any>();
    this.ChangedRoleByRelationType.set(roleType.relationType, role);
    this.OnChange();
  }

  protected get Id(): number {
    return this.Strategy.id;
  }

  protected get Class(): Class {
    return this.Strategy.class;
  }

  protected get Session(): Session {
    return this.Strategy.Session;
  }

  protected get Workspace(): Workspace {
    return this.Strategy.Workspace;
  }
}
